src/Avalonia.Base/Input/AccessKeyHandler.cs

```diff
@@ -255,5 +313,277 @@ private void MainMenuClosed(object? sender, EventArgs e)
+            }
+
+            return finalTargets;
+        }
+
+        private static bool IsTargetable(IInputElement element) =>
+            element is { IsEffectivelyEnabled: true, IsEffectivelyVisible: true };
+
+        private List<IInputElement> CopyAndPurgeDead(string key)
+        {
+            lock (_registrations)
+            {
+                var registrationsToRemove = _registrations
+                    .Where(m => m.GetInputElement() == null)
+                    .ToList();
+
+                foreach (var registration in registrationsToRemove)
+                {
+                    _registrations.Remove(registration);
+                }
+
+                return _registrations
+                    .Where(m => m.Key == key)
+                    .Select(m => m.GetInputElement())
+                    .OfType<IInputElement>()
+                    .ToList();
+            }
+        }
+
+        /// <summary>
+        /// Returns targeting information for the given element.
+        /// </summary>
+        /// <param name="element"></param>
+        /// <param name="key"></param>
+        /// <returns>AccessKeyInformation with target for the access key.</returns>
+        private static AccessKeyInformation GetInfoForElement(IInputElement? element, string key)
+        {
+            var info = new AccessKeyInformation();
+            if (element == null)
+                return info;
+
+            var args = new AccessKeyPressedEventArgs(key);
+            element.RaiseEvent(args);
+            info.Target = args.Target;
+
+            return info;
+        }
+
+        /// <summary>
+        /// Checks if the focused element is a descendent of the owner.
+        /// </summary>
+        /// <param name="owner">The owner to check.</param>
+        /// <returns>If focused element is decendant of owner <c>true</c>, otherwise <c>false</c>. </returns>
+        private static bool IsFocusWithinOwner(IInputRoot owner)
+        {
+            var focusedElement = KeyboardDevice.Instance?.FocusedElement;
+            if (focusedElement is not InputElement inputElement)
+                return false;
+
+            var isAncestorOf = owner is Visual root && root.IsVisualAncestorOf(inputElement);
+            return isAncestorOf;
+        }
+
+        /// <summary>
+        /// Sorts the list of targets in hierarchical order to ensure that elements within a tab,
+        /// for example, are processed before the next tab item in case of identical access keys
+        /// </summary>
+        private static List<IInputElement> SortByHierarchy(List<IInputElement> targets)
+        {
+            var sorted = new List<IInputElement>();
+            var elements = targets.OfType<InputElement>().ToList();
+            for (var i = 0; i < elements.Count; i++)
+            {
+                var parent = elements[i];
+                if (sorted.Contains(parent))
+                    continue;
+
+                sorted.Add(parent);
```

Comment:
I simplified the sorting code for clarity, see [here](https://github.com/royalapplications/Avalonia/blob/36b7e77dc81c3fe76ff618528b3cfc498274ce4b/src/Avalonia.Base/Input/AccessKeyHandler.cs#L457), but basically it is the same procedure.
The original order of the incoming elements is the registration order, but in case of nested elements, for example in tabs, I want to run over the child elements before going to the next parent. 
The tabs on the Accelerator page in the catalog show the behavior, it feels quite intuitive.